app/bridge/client.js

```diff
@@ -90,85 +121,142 @@ function createNewClient(name, version) {
     this.client = client;
     this.uuid = uuid();
 
-    this.server = null;
+    this.serverChannel = null;
+
+    this.connected = false;
+    this.connectionDeferred = null;
+    this.disconnectionDeferred = null;
```

Comment:
The way those are used in the code is a bit confusing. Can you instead have a kind of state for the client. Something like:

var kStates = {
  'Disconnected': 0,
  'Connecting': 1,
  'Connected': 2,
  'Disconnecting': 3
}

this.state = kStates.Disconnected;

And then use that in a

switch(this.state) {
  case kStates.Disconnected:
    break;

 case kStates.Connecting:
   break;

  case kStates.Connected:
    break;

  case kStates.Disconnecting:
    break;

  default:
    Throw new Error('Unsupported state: ' + this.state);
   break;
} 

Then you can have a separate property for the running promise as I assume it will be one per state.